drivers/sensor/bq34110/bq34110.c

```diff
@@ -0,0 +1,373 @@
+		{ REMAINING_CAPACITY, &bq34110->remaining_charge_capacity },
+	};
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+	for (size_t i = 0; i < ARRAY_SIZE(regs); i++) {
+		int rc;
+
+		rc = bq34110_cmd_reg_read(bq34110, regs[i].reg_addr,
+					  regs[i].dest);
+		if (rc != 0) {
+			LOG_ERR("Failed to read channel %d", chan);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int bq34110_ctrl_reg_read(struct bq34110_data *bq34110,
+				 uint16_t subcommand,
+				 uint16_t *val, uint8_t read_addr)
+{
+	int ret;
+	uint8_t i2c_write_data[3], read_data[2], read_reg;
+
+	memset(&i2c_write_data, 0, sizeof(i2c_write_data));
+	i2c_write_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	i2c_write_data[1] = (uint8_t)(subcommand & 0x00FF);
+	i2c_write_data[2] = (uint8_t)((subcommand >> 8) & 0x00FF);
+	ret = i2c_write(bq34110->i2c, i2c_write_data, 3,
+			DT_INST_REG_ADDR(0));
+	if (ret < 0) {
+		LOG_ERR("Failed to write");
+		return ret;
+	}
+
+	k_msleep(BQ34110_DELAY);
+
+	read_reg = read_addr;
+	ret = i2c_write_read(bq34110->i2c, DT_INST_REG_ADDR(0), &read_reg,
+			     1, &read_data, 2);
+	if (ret < 0) {
+		LOG_ERR("Failed to read write");
+		return ret;
+	}
+
+	*val = ((read_data[1] << 8) | read_data[0]);
+
+	return 0;
+}
+
+static int updateDFparameter(struct bq34110_data *bq34110, uint16_t df_addr,
+			     uint8_t data[], uint8_t datalength)
+{
+	int ret;
+	uint8_t i2c_df_addr_data[3], i2c_df_write_data[datalength+1], i;
+	uint8_t i2c_datasum_datalen[3];
+	uint16_t datasum;
+
+	i2c_df_addr_data[0] = MANUFACTURER_ACCESS_CONTROL;
+	i2c_df_addr_data[1] = (uint8_t)(df_addr & 0x00FF);
+	i2c_df_addr_data[2] = (uint8_t)((df_addr >> 8) & 0x00FF);
+
+	ret = i2c_write(bq34110->i2c, i2c_df_addr_data,
+			sizeof(i2c_df_addr_data), DT_INST_REG_ADDR(0));
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	i2c_df_write_data[0] = MAC_DATA;
+	for (i = 1; i <= datalength; i++) {
+		i2c_df_write_data[i] = data[i-1];
+	}
+
+	ret = i2c_write(bq34110->i2c, i2c_df_write_data,
+			sizeof(i2c_df_write_data), DT_INST_REG_ADDR(0));
+	if (ret < 0) {
+		LOG_ERR("Failed to write MAC");
+		return ret;
+	}
+
+	datasum = i2c_df_addr_data[1] + i2c_df_addr_data[2];
+	for (i = 1; i <= datalength; i++) {
+		datasum += i2c_df_write_data[i];
+	}
+
+	i2c_datasum_datalen[0] = MAC_DATA_SUM;
+	i2c_datasum_datalen[1] = (uint8_t)~datasum;
+	i2c_datasum_datalen[2] = 4 + datalength;
+	ret = i2c_write(bq34110->i2c, i2c_datasum_datalen,
+			3, DT_INST_REG_ADDR(0));
+	if (ret < 0) {
+		LOG_ERR("Failed to write MACDatasum");
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief initialise the fuel gauge
+ *
+ * @return 0 for success
+ */
+static int bq34110_gauge_init(const struct device *dev)
+{
+	int ret;
+	uint8_t writedata[2];
+	int16_t design_capacity;
+	struct bq34110_data *bq34110 = dev->data;
+	const struct bq34110_config *const config = dev->config;
+	uint16_t design_voltage, device_type, taper_current, taper_voltage;
+
+	bq34110->i2c = device_get_binding(config->bus_name);
```

Comment:
Please use DEVICE_DT_GET(DT_INST_BUS(index)) in the BQ34110_INIT macro to get the bus device directly and use device_is_ready to check that the pointer is valid. See the BME280 driver for an example.

Reply:
@mbolivar-nordic fixed.